tests/ga/test_signature_validation.py

```diff
@@ -0,0 +1,113 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+# Requires Python 2.6+ and Openssl 1.0+
+#
+import os
+import sys
+
+from tests.lib.tools import AgentTestCase, data_dir, patch, skip_if_predicate_true
+from azurelinuxagent.ga.signing_certificate_util import write_signing_certificates
+from azurelinuxagent.ga.signature_validation import validate_signature, SignatureValidationError
+from azurelinuxagent.common.event import WALAEventOperation
+
+
+class TestSignatureValidation(AgentTestCase):
+    def setUp(self):
+        AgentTestCase.setUp(self)
+        write_signing_certificates()
+        self.null_ext_zip_path = os.path.join(data_dir, "signing/null_extension.zip")
+        null_ext_sig_path = os.path.join(data_dir, "signing/null_extension_signature.txt")
+        with open(null_ext_sig_path, 'r') as f:
+            self.null_ext_signature = f.read()
+
+    def tearDown(self):
+        patch.stopall()
+        AgentTestCase.tearDown(self)
+
+    def test_should_validate_signature_successfully(self):
+        """
+        Test that the signature can be validated successfully without raising an exception.
+
+        Note: The test extension was signed with an expired leaf certificate (expired 2024), but it
+        should still validate successfully because the signature was generated when all certs were unexpired.
+        Currently, we do not have a version of NullExtension signed with an unexpired cert. While we could request a
+        newly signed version, leaf certs expire fairly quickly (within a year) and we would need to frequently update
+        the test with a new signature and package.
+        """
+        validate_signature(self.null_ext_zip_path, self.null_ext_signature)
+
+    def test_should_raise_error_if_signature_does_not_match_package(self):
+        # This signature is correctly formatted but belongs to a different extension (CSE),
+        # signature validation should fail for null extension
+        with open(os.path.join(data_dir, "signing/invalid_signature.txt"), 'r') as f:
+            invalid_signature = f.read()
+            with self.assertRaises(SignatureValidationError, msg="Signature is invalid, should have raised error"):
+                validate_signature(self.null_ext_zip_path, invalid_signature)
+
+    def test_should_raise_error_if_package_is_tampered_with(self):
+        # This is the null extension zip package with one byte modified, signature validation should fail
+        modified_ext = os.path.join(data_dir, "signing/null_extension_modified.zip")
+        with self.assertRaises(SignatureValidationError, msg="Zip package does not match signature, should have raised error"):
+            validate_signature(modified_ext, self.null_ext_signature)
+
+    def test_should_raise_error_on_incorrect_signing_certificate(self):
+        # This certificate (unexpired) does not match the one used for signing - signature validation should fail
```

Comment:
is this test certificate coming from the signing authority, or something you generated?